File: references/betterlectio/ios/BetterLectio/ReviewEligibility.swift
import Foundation

enum ReviewTrigger: String, Sendable {
    case homeworkDone = "homework_done"
    case scheduleLoaded = "schedule_loaded"
    case privateEventCreated = "private_event_created"
    case messageSent = "message_sent"

    var analyticsName: String { rawValue }
}

/// Pure eligibility checks for the soft rating pre-filter.
/// Mirrors Android `ReviewEligibility`.
enum ReviewEligibility {
    static let minInstallAgeDays = 14
    static let minLaunchesInWindow = 8
    static let promptHourStart = 16
    static let promptHourEnd = 21
    static let cooldownDays = 90
    static let maxLifetimePrompts = 3
    static let minSessionSeconds: TimeInterval = 30
    static let errorCooldownSeconds: TimeInterval = 60

    /// Mon–Thu only (same as Android).
    static let promptWeekdays: Set<Int> = [
        2, // Monday
        3, // Tuesday
        4, // Wednesday
        5, // Thursday
    ]

    static func isEligible(
        store: ReviewPromptStore = .shared,
        now: Date = Date(),
        calendar: Calendar = .current,
        sessionStartedAt: Date?,
        lastErrorAt: Date?
    ) -> Bool {
        if store.neverAsk() || store.completedStoreFlow() { return false }
        if store.promptCount() >= maxLifetimePrompts { return false }
        if !installedLongEnough(now: now) { return false }
        if store.launchCountInWindow(now: now) < minLaunchesInWindow { return false }
        if !inPromptWindow(now: now, calendar: calendar) { return false }
        if !cooldownElapsed(store: store, now: now) { return false }
        if !sessionCalm(now: now, sessionStartedAt: sessionStartedAt, lastErrorAt: lastErrorAt) {
            return false
        }
        return true
    }

    static func installedLongEnough(now: Date = Date()) -> Bool {
        guard let installedAt = firstInstallDate() else { return false }
        let days = Calendar.current.dateComponents([.day], from: installedAt, to: now).day ?? 0
        return days >= minInstallAgeDays
    }

    /// Approximate first-install time via the app documents directory creation date.
    static func firstInstallDate() -> Date? {
        guard
            let docs = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first,
            let attrs = try? FileManager.default.attributesOfItem(atPath: docs.path),
            let created = attrs[.creationDate] as? Date
        else {
            return nil
        }
        return created
    }

    private static func inPromptWindow(now: Date, calendar: Calendar) -> Bool {
        let weekday = calendar.component(.weekday, from: now)
        guard promptWeekdays.contains(weekday) else { return false }
        let hour = calendar.component(.hour, from: now)
        return hour >= promptHourStart && hour < promptHourEnd
    }

    private static func cooldownElapsed(store: ReviewPromptStore, now: Date) -> Bool {
        guard let last = store.lastPromptAt() else { return true }
        let days = Calendar.current.dateComponents([.day], from: last, to: now).day ?? 0
        return days >= cooldownDays
    }

    private static func sessionCalm(
        now: Date,
        sessionStartedAt: Date?,
        lastErrorAt: Date?
    ) -> Bool {
        guard let sessionStartedAt else { return false }
        if now.timeIntervalSince(sessionStartedAt) < minSessionSeconds { return false }
        if let lastErrorAt, now.timeIntervalSince(lastErrorAt) < errorCooldownSeconds {
            return false
        }
        return true
    }
}

File: android/app/src/main/java/dk/betterw4/android/feature/review/ReviewEligibility.kt
package dk.betterw4.android.feature.review

import android.content.Context
import android.content.pm.PackageManager
import dagger.hilt.android.qualifiers.ApplicationContext
import java.time.DayOfWeek
import java.time.Instant
import java.time.LocalDate
import java.time.LocalTime
import java.time.ZoneId
import java.time.temporal.ChronoUnit
import javax.inject.Inject
import javax.inject.Singleton

enum class ReviewTrigger(val analyticsName: String) {
    HomeworkDone("homework_done"),
    ScheduleLoaded("schedule_loaded"),
    PrivateEventCreated("private_event_created"),
    MessageSent("message_sent"),
}

/**
 * Pure eligibility checks for the soft rating pre-filter.
 */
@Singleton
class ReviewEligibility @Inject constructor(
    @ApplicationContext private val context: Context,
    private val store: ReviewPromptStore,
) {
    fun isEligible(
        now: Instant = Instant.now(),
        zone: ZoneId = ZoneId.systemDefault(),
        sessionStartedAt: Instant?,
        lastErrorAt: Instant?,
    ): Boolean {
        if (store.neverAsk() || store.completedPlayFlow()) return false
        if (store.promptCount() >= MAX_LIFETIME_PROMPTS) return false
        if (!installedLongEnough(now)) return false
        if (store.launchCountInWindow(now) < MIN_LAUNCHES_IN_WINDOW) {
            return false
        }
        if (!inPromptWindow(LocalTime.ofInstant(now, zone), LocalDate.ofInstant(now, zone).dayOfWeek)) {
            return false
        }
        if (!cooldownElapsed(now)) return false
        if (!sessionCalm(now, sessionStartedAt, lastErrorAt)) return false
        return true
    }

    fun installedLongEnough(now: Instant = Instant.now()): Boolean {
        val installedAt = firstInstallInstant() ?: return false
        return ChronoUnit.DAYS.between(installedAt, now) >= MIN_INSTALL_AGE_DAYS
    }

    private fun firstInstallInstant(): Instant? {
        return try {
            val info = context.packageManager.getPackageInfo(context.packageName, 0)
            Instant.ofEpochMilli(info.firstInstallTime)
        } catch (_: PackageManager.NameNotFoundException) {
            null
        }
    }

    private fun inPromptWindow(time: LocalTime, day: DayOfWeek): Boolean {
        if (day !in PROMPT_WEEKDAYS) return false
        val hour = time.hour
        return hour in PROMPT_HOUR_START until PROMPT_HOUR_END
    }

    private fun cooldownElapsed(now: Instant): Boolean {
        val last = store.lastPromptAtMillis()
        if (last <= 0L) return true
        val elapsedDays = ChronoUnit.DAYS.between(Instant.ofEpochMilli(last), now)
        return elapsedDays >= COOLDOWN_DAYS
    }

    private fun sessionCalm(
        now: Instant,
        sessionStartedAt: Instant?,
        lastErrorAt: Instant?,
    ): Boolean {
        if (sessionStartedAt == null) return false
        if (ChronoUnit.SECONDS.between(sessionStartedAt, now) < MIN_SESSION_SECONDS) return false
        if (lastErrorAt != null &&
            ChronoUnit.SECONDS.between(lastErrorAt, now) < ERROR_COOLDOWN_SECONDS
        ) {
            return false
        }
        return true
    }

    companion object {
        const val MIN_INSTALL_AGE_DAYS = 14L
        const val MIN_LAUNCHES_IN_WINDOW = 8
        const val PROMPT_HOUR_START = 16
        const val PROMPT_HOUR_END = 21
        const val COOLDOWN_DAYS = 90L
        const val MAX_LIFETIME_PROMPTS = 3
        const val MIN_SESSION_SECONDS = 30L
        const val ERROR_COOLDOWN_SECONDS = 60L
        val PROMPT_WEEKDAYS = setOf(
            DayOfWeek.MONDAY,
            DayOfWeek.TUESDAY,
            DayOfWeek.WEDNESDAY,
            DayOfWeek.THURSDAY,
        )
    }
}
